src/gin/nccl_ofi_gin.cpp

```diff
@@ -0,0 +1,734 @@
+	    signalOp != NCCL_NET_SIGNAL_OP_ADD) {
+		NCCL_OFI_WARN("Only support signal add/increment");
+		return -EINVAL;
+	}
+
+	auto *rdma_ep = reinterpret_cast<nccl_net_ofi_rdma_ep_t *>(gin_comm->ep);
+	auto &rank_comm = gin_comm->rank_comms[rank];
+	uint16_t msg_seq_num = rank_comm.next_target_seq_num;
+	uint32_t remote_comm_id = rank_comm.comm_id;
+	uint16_t rail_id = gin_comm->next_rail_id;
+	gin_comm->next_rail_id = (gin_comm->next_rail_id + 1) % rdma_ep->num_rails;
+
+	int ret = 0;
+	nccl_net_ofi_gin_tx_req_t *write_req = nullptr;
+
+	if (size > 0) {
+		write_req = new nccl_net_ofi_gin_tx_req_t();
+
+		void *src = reinterpret_cast<void *>(srcMhandle->remote_mr[gin_comm->rank].address + srcOff);
+		auto *src_mhandle = static_cast<nccl_net_ofi_rdma_mr_handle_t *>(srcMhandle->local_comm_handle);
+		void *desc = fi_mr_desc(src_mhandle->mr[rail_id].get());
+
+		auto &dest_remote_mr = dstMhandle->remote_mr[rank];
+		uint64_t dest = dest_remote_mr.address + dstOff;
+		uint64_t data = GET_RDMA_WRITE_IMM_DATA(remote_comm_id, msg_seq_num, 1);
+
+		auto op = [=]() {
+			ssize_t rc = fi_writedata(rdma_ep->rails[rail_id].ofi_ep.get(), src, size, desc, data,
+						  rank_comm.address[rail_id], dest,
+						  dest_remote_mr.mr_key[rail_id],
+						  &(write_req->ctx.ofi_ctx));
+
+			if (rc != 0 && rc != -FI_EAGAIN) {
+				NCCL_OFI_WARN("Failed call to fi_writedata; RC: %zd", rc);
+			}
+			return rc;
+		};
+
+		ret = op();
+		if (ret == -FI_EAGAIN) {
+			gin_comm->pending_requests.push_back(op);
+			ret = 0;
+		} else if (ret != 0) {
+			delete write_req;
+			return ret;
+		}
+	}
+
+	nccl_ofi_freelist_elem_t *metadata_elem =
+		nccl_ofi_freelist_entry_alloc(gin_comm->metadata_fl);
+	if (!metadata_elem) {
+		delete write_req;
+		return -ENOMEM;
+	}
+
+	auto *metadata_send = static_cast<nccl_net_ofi_rdma_signal_metadata_msg_t *>
+		(metadata_elem->ptr);
+	freelist_regmr_fn_handle_t *fl_handle =
+		(freelist_regmr_fn_handle_t *)metadata_elem->mr_handle;
+	metadata_send->type = NCCL_OFI_RDMA_MSG_GIN_METADATA;
+	metadata_send->msg_seq_num = msg_seq_num;
+	metadata_send->num_write_segments = (size > 0) ? 1 : 0;
+	metadata_send->remote_comm_id = remote_comm_id;
+	metadata_send->signal_base_address = (signalMhandle ? signalMhandle->remote_mr[rank].address
+						: 0);
+	metadata_send->signal_offset = signalOff;
+	metadata_send->signal_value = signalValue;
+
+	nccl_net_ofi_gin_tx_req_t *send_req = new nccl_net_ofi_gin_tx_req_t();
+
+	auto op = [=]() {
+		ssize_t rc = fi_send(rdma_ep->rails[rail_id].ofi_ep.get(), metadata_send, sizeof(*metadata_send),
+				     fi_mr_desc(fl_handle->mr_handle->mr[rail_id].get()), rank_comm.address[rail_id],
+				     &send_req->ctx.ofi_ctx);
+		if (rc != 0 && rc != -FI_EAGAIN) {
+			NCCL_OFI_WARN("fi_send failed with RC %zd", rc);
+		}
+
+		return rc;
+	};
+	ret = op();
+	if (ret == -FI_EAGAIN) {
+		gin_comm->pending_requests.push_back(op);
+		ret = 0;
+	} else if (ret != 0) {
+		delete write_req;
```

Comment:
Can you delete `write_req` here given that it was conditionally created `if (size > 0)` before this?